src/main/scala/com/github/krasserm/ases/EventSourcing.scala

```diff
@@ -227,9 +228,9 @@ private class EventSourcing[S, E, REQ, RES](
         tryPullEi()
 
       private def tryPullEi(): Unit =
-        if (!requestUpstreamFinished) pull(ei)
+        if (!requestUpstreamFinished && !hasBeenPulled(ei) && !isClosed(ei)) pull(ei)
 
       private def tryPullCi(): Unit =
-        if (isAvailable(eo) && isAvailable(ro) && !hasBeenPulled(ci) && roundtrip.isEmpty && !requestUpstreamFinished) pull(ci)
+        if (!requestUpstreamFinished && recovered && roundtrip.isEmpty && isAvailable(eo) && isAvailable(ro) && !hasBeenPulled(ci) && !isClosed(ci)) pull(ci)
```

Comment:
`!isClosed(ci)` is redundant with `!requestUpstreamFinished` and can be removed.